org.widelands.widelands.json

```diff
@@ -0,0 +1,478 @@
+                        if [[ -n \"${PATH}\" ]]; then
+                            PATH=\"${PATH}:${xmlstarlet_bin}\";
+                        else
+                            PATH=\"${xmlstarlet_bin}\";
+                        fi;
+                        export PATH;
+                        xmlstarlet ed --inplace -d '/component/releases/release[position()>1]' \"${app_name}.appdata.xml\";
+                        echo;"
+                    ],
+                    "dest-filename": "flatpak_edit_appdata"
+                },
+                {
+                    "type": "script",
+                    "commands": [
+                        "cd \"$( dirname \"$( readlink -f \"${0}\" )\" )\";",
+                        "source './flatpak_setup';",
+                        "echo 'install icons';
+                        find 'icons' -mindepth 2 -maxdepth 2 -type f -name \"${app_name}.png\" | sort -V | xargs -I{} dirname '{}' | xargs -I{} basename '{}' | while read -r size ; do
+                            install -p -D -m 0644 \"icons/${size}/${app_name}.png\" \"/app/share/icons/hicolor/${size}/apps/${app_name}.png\";
+                        done;
+                        echo;"
+                    ],
+                    "dest-filename": "flatpak_install_icons"
+                },
+                {
+                    "type": "script",
+                    "commands": [
+                        "cd \"$( dirname \"$( readlink -f \"${0}\" )\" )\";",
+                        "source './flatpak_setup';",
+                        "echo 'install pixmaps';
+                        install -d \"/app/share/pixmaps\";
+                        ln -s \"../${app_name}/images/logos/wl-logo-64.png\" \"/app/share/pixmaps/${app_name}.png\";
+                        echo;"
+                    ],
+                    "dest-filename": "flatpak_install_pixmaps"
+                },
+                {
+                    "type": "script",
+                    "commands": [
+                        "cd \"$( dirname \"$( readlink -f \"${0}\" )\" )\";",
+                        "source './flatpak_setup';",
+                        "echo 'install desktop';
+                        install -p -D -m 0644 \"${app_name}.desktop\" \"/app/share/applications/${app_name}.desktop\";
+                        echo;"
+                    ],
+                    "dest-filename": "flatpak_install_desktop"
+                },
+                {
+                    "type": "script",
+                    "commands": [
+                        "cd \"$( dirname \"$( readlink -f \"${0}\" )\" )\";",
+                        "source './flatpak_setup';",
+                        "echo 'install appdata';
+                        install -p -D -m 0644 \"${app_name}.appdata.xml\" \"/app/share/metainfo/${app_name}.appdata.xml\";
+                        echo;"
+                    ],
+                    "dest-filename": "flatpak_install_appdata"
```

Comment:
I really dislike this layout. You replaced literally 1 line of logic with 12...

Reply:
> I really dislike this layout. You replaced literally 1 line of logic with 12...

I really don't know what you expect. You didn't like the previous layout that I used in retext, so I split the code into separate parts. You didn't like it either.
There are things that have to be (or at least should be) done *directly* in the flatpak manifest, like icon scaling, editing the desktop/appdata file, and installing the necessary files - just like in a RPM SPEC. You can't just patch an icon or AppData XML file (quoting from the Bash FAQ: _You can't parse structured languages like HTML or XML using bash, grep, sed, cut, etc. Look into xmllint, xmlstarlet, xpath, xsltproc_). Moreover, as you probably already know, the Flatpak's SDKs lack many useful tools for creating packages, like ImageMagick, XML tools (xmlstarlet, xsltproc, etc), text encoding tools/format converters (dos2unix), PO tools (po4a), and the actions mentioned above heavily depend on these programs. IMO creating a truly separated "post-install" script to do this is just stupid, because it will depend on tools from modules created directly in the flatpak manifest (see: LD_LIBRARY_PATH, PATH), and that should be purged after built is done. This will only reduce the clarity of the code.

The actions that must be taken for this particular package are:
- icon resize: one simple loop using the `convert` command from ImageMagick
- edit desktop file: change one entry using the `desktop-file-edit` tool
- edit AppData file using the `xmlstarlet` program
- install necessary files into proper locations using the `install` command

These things have to be done in one way or another: either using separate "scripts" in flatpak manifest (just like now) or in one wall of commands in the `post-install` section. You choose.

Reply:
Just have a single script, I fail to see why adding complication on that is an improvement.

Reply:
BTW to convert an svg to a png you can just use `rsvg-convert` which is in the Sdk.